b2g/components/RemoteControlService.jsm

```diff
@@ -491,30 +624,342 @@ function getCommandJSSandbox(aImportFunctions) {
+  this._client_round2 = null;
 }
 EventHandler.prototype = {
   // PUBLIC FUNCTIONS
   handleEvent: function(aEvent) {
-    // TODO: Implement JPAKE pairing (Bug 1207996)
     switch (aEvent.type) {
-      // Implement control command dispatch
+      case "auth":
+        // Implement JPAKE authentication
+        this._handleAuthEvent(aEvent);
+        break;
       case "command":
+        // Implement control command dispatch
         this._handleCommandEvent(aEvent);
         break;
       default:
+        // Not accepted event type, report error to client
+        this._handleError(this, "common", "Not accepted event type");
+        break;
+    }
+  },
+
+  _handleAuthEvent: function(aEvent) {
+    switch (aEvent.action) {
+      case "request_handshake":
+        this._handleHandshake(aEvent);
+        break;
+      case "jpake_client_1":
+        this._handleJPAKERound1Event(aEvent);
+        break;
+      case "jpake_client_2":
+        this._handleJPAKERound2Event(aEvent);
+        break;
+      case "client_key_confirmation":
+        this._handleKeyConfirmEvent(aEvent);
+        break;
+      default:
         break;
     }
   },
 
-  // Pass and run command event in sandbox
+  _handleHandshake: function(aEvent) {
+    if (this._status != EVENT_HANDLER_STATUS.INITIAL) {
+      DEBUG && debug("Status not in INITIAL, drop the event");
+      this._handleError(this, aEvent.type, "Wrong status");
+      return;
+    }
+
+    let reply = {
+      type: "auth",
+      action: "response_handshake",
+      detail: 1
+    };
+
+    // If client sends an existing ID, then reply is 2 (second connection)
+    // Also restore AES and HMAC key from UUID
+    if (aEvent.detail && aEvent.detail.id && this._server.getKeysFromUUID(aEvent.detail.id) !== null) {
+      reply.detail = 2;
+
+      this._clientID = aEvent.detail.id;
+      let keys = this._server.getKeysFromUUID(aEvent.detail.id);
+      this._aes256B64 = { value: keys.aesKey };
+      this._hmac256B64 = { value: keys.hmacKey };
+    }
+
+    this._connection.sendMsg(reply);
+    this._status = EVENT_HANDLER_STATUS.ROUND_1;
+  },
+
+  _handleJPAKERound1Event: function(aEvent) {
+    if (this._status != EVENT_HANDLER_STATUS.ROUND_1) {
+      DEBUG && debug("Status not in ROUND_1, drop the event");
+      this._handleError(this, aEvent.type, "Wrong status");
+      return;
+    }
+
+    // First time connection, generate PIN, trigger notification on screen
+    if (this._clientID === null) {
+      let pin = this._server.getPIN();
+      if(pin === null) {
+        pin = this._server.generatePIN();
+        // Show notification on screen
+        SystemAppProxy._sendCustomEvent(REMOTE_CONTROL_EVENT, { pincode: pin, action: 'pin-created' });
+      }
+    }
+
+    // Store client round 1
+    this._client_round1 = aEvent.detail;
+
+    // Compute server round 1
+    this._JPAKE = Cc["@mozilla.org/services-crypto/sync-jpake;1"]
+                    .createInstance(Ci.nsISyncJPAKE);
```

Comment:
Unfortunately this uses a prime group. It would be better if it used ECC, but that's probably nothing we can change here. But important to keep in mind.